### DISCUSSION TITLE
Enforce "primaryKey" when "attributes" is defined

### DISCUSSION THREAD

**Discussion by JamieS1211** [sequelize/sequelize#15843] [Help & Questions]
Hi there,

When "attributes" is defined in the "FindOptions" as expected all other attributes are undefined, including the "primaryKey". This is problematic for: `model.reload()` and Includes.

If you fetch a model, specifying the "attributes" but omitting the "primaryKey" there seems to be no way to recover the original model, with the `model.reload()` function throwing a "WHERE parameter "id" has invalid "undefined" value" error.

In terms of "includes" to prevent this scenario causing an error, in the code for model there is a specific addition of "primaryKey" in the "attributes" (not "originalAttributes") to ensure that the "includes" succeeds. This affects the resultant database query, however sadly this is not accessible via the model as its not in the "originalAttributes" value.

Is there any setting or recommended way to enforce that the "primaryKey" is always fetched from the database and available on the model when fetching data? Should this be done via scopes? Should it be done via the "beforeFindAfterOptions" hook? Any advice here would be great.

**Comment by ephys** [comment]
There isn't really a universal solution for this. `beforeFind` could work

In Sequelize 7, we're printing an explicit error when using a method that needs the primary key, but the PK is missing: https://github.com/sequelize/sequelize/blob/3cfe6722cee3dc81ba5582dfddbddae5f9bc9c53/packages/core/src/model.js#L3115

**Comment by JamieS1211** [comment]
Thanks for the reply.

I actually tried with the `beforeFind` hook however, as it only runs on the root level so if you use eager fetching with "includes" you can still find yourself in a problem.


        const itemA = await ModelA.findOne({ include: [{ model: ModelB, attributes: ["fieldB"] }] })
        assertIsDefined(itemA)

        await itemA.itemB.reload()


In the above example assuming fieldB is not the primaryKey, the `reload()` will fail
